cub/cub/device/dispatch/dispatch_transform.cuh

```diff
@@ -0,0 +1,922 @@
+  const Offset offset = static_cast<Offset>(blockIdx.x) * tile_size;
+
+  [&](cuda::std::array<value_t<RandomAccessIteartorIn>, ItemsPerThread>&&... arrays) {
+  // Load ItemsPerThread times
+#pragma unroll
+    for (int j = 0; j < ItemsPerThread; ++j)
+    {
+      const auto idx = offset + (j * BlockDim + threadIdx.x);
+      if (idx < len)
+      {
+        // TODO(bgruber): replace by fold over comma in C++17
+        int dummy[] = {(arrays[j] = ins[idx], 0)..., 0}; // extra zero to handle empty packs
+        (void) dummy;
+      }
+    }
+    // Execute ItemsPerThread times
+#pragma unroll
+    for (int j = 0; j < ItemsPerThread; ++j)
+    {
+      const auto idx = offset + (j * BlockDim + threadIdx.x);
+      if (idx < len)
+      {
+        out[idx] = f(arrays[j]...);
+      }
+    }
+  }(cuda::std::array<value_t<RandomAccessIteartorIn>, ItemsPerThread>{}...);
+}
+
+// TODO(bgruber) cheap copy of ::cuda::std::apply, which requires C++17.
+template <class F, class Tuple, std::size_t... Is>
+_CCCL_DEVICE auto poor_apply_impl(F&& f, Tuple&& t, ::cuda::std::index_sequence<Is...>)
+  -> decltype(std::forward<F>(f)(::cuda::std::get<Is>(std::forward<Tuple>(t))...))
+{
+  return std::forward<F>(f)(::cuda::std::get<Is>(std::forward<Tuple>(t))...);
+}
+template <class F, class Tuple>
+_CCCL_DEVICE auto poor_apply(F&& f, Tuple&& t)
+  -> decltype(poor_apply_impl(
+    std::forward<F>(f),
+    std::forward<Tuple>(t),
+    ::cuda::std::make_index_sequence<::cuda::std::tuple_size<::cuda::std::__libcpp_remove_reference_t<Tuple>>::value>{}))
+{
+  return poor_apply_impl(
+    std::forward<F>(f),
+    std::forward<Tuple>(t),
+    ::cuda::std::make_index_sequence<::cuda::std::tuple_size<::cuda::std::__libcpp_remove_reference_t<Tuple>>::value>{});
+}
+
+// mult must be a power of 2
+_CCCL_HOST_DEVICE inline auto round_up_to_multiple(int x, int mult) -> int
+{
+  const auto p = cuda::std::popcount(static_cast<unsigned>(mult));
+  assert(p == 1);
+  return (x + mult - 1) & ~(mult - 1);
+}
+
+// TODO(bgruber): inline this as lambda in C++14
+template <typename T>
+_CCCL_DEVICE T* copy_and_return_smem_dst(
+  cooperative_groups::thread_block& group, int tile_size, char* smem, int& smem_offset, int global_offset, const T* ptr)
+{
+  // using T          = ::cuda::std::__remove_const_t<::cuda::std::__remove_pointer_t<decltype(ptr)>>;
+  const auto count = static_cast<uint32_t>(sizeof(T)) * tile_size;
+  smem_offset      = round_up_to_multiple(smem_offset, alignof(T));
+  auto smem_dst    = reinterpret_cast<T*>(smem + smem_offset);
+  cooperative_groups::memcpy_async(group, smem_dst, ptr + global_offset, count);
+  smem_offset += count;
+  return smem_dst;
+}
+
+template <typename Offset, typename F, typename RandomAccessIteartorOut, typename... InTs>
+CUB_DETAIL_KERNEL_ATTRIBUTES void transform_memcpy_async_kernel(
+  Offset len, int num_elem_per_thread, F f, RandomAccessIteartorOut out, const InTs*... pointers)
+{
+#if CUB_PTX_ARCH >= 800
```

Comment:
It's  indeed not necessary. I thought it may, since for `ublkcp` I need it, so maybe for `cooperative_groups::memcpy_async` as well. But the latter also compiles on previous architecturs. I removed the `#if` now.